prog/vnet/nic_nexus.rb

```diff
@@ -0,0 +1,109 @@
+# frozen_string_literal: true
+
+class Prog::Vnet::NicNexus < Prog::Base
+  semaphore :destroy, :refresh_mesh, :detach_vm
+
+  def self.assemble(project_id, private_subnet_id, name: nil, ipv6_addr: nil, ipv4_addr: nil)
+    project = Project[project_id]
+    unless project || Config.development?
+      fail "No existing project"
+    end
+
+    # maybe wait vnet
+    subnet = PrivateSubnet[private_subnet_id]
+    unless subnet
+      fail "Given subnet doesn't exist with the id #{private_subnet_id}"
+    end
+
+    ubid = UBID.generate(UBID::TYPE_NIC)
+    tap_name = Nic.ubid_to_tap_name(ubid)
+
+    Validation.validate_name(tap_name)
+
+    ipv6_addr ||= subnet.random_private_ipv6.to_s
+    ipv4_addr ||= subnet.random_private_ipv4.to_s
+
+    DB.transaction do
+      nic = Nic.create(private_ipv6: ipv6_addr, private_ipv4: ipv4_addr, mac: gen_mac,
+        encryption_key: gen_encryption_key, name: name, tap_name: tap_name,
+        private_subnet_id: private_subnet_id) { _1.id = ubid.to_uuid }
+      subnet.add_nic(nic)
+      Strand.create(prog: "Vnet::NicNexus", label: "wait_vnet") { _1.id = ubid.to_uuid }
+    end
+  end
+
+  def self.gen_encryption_key
+    "0x" + SecureRandom.bytes(36).unpack1("H*")
+  end
+
+  def nic
+    @nic ||= Nic[strand.id]
+  end
+
+  def wait_vnet
+    unless nic.private_subnet.strand.label == "wait"
+      nap 1
+    end
+
+    hop :wait
+  end
+
+  def wait
+    when_destroy_set? do
+      hop :destroy
+    end
+
+    when_refresh_mesh_set? do
+      hop :refresh_mesh
+    end
+
+    when_detach_vm_set? do
+      hop :detach_vm
+    end
+
+    nap 30
+  end
+
+  def refresh_mesh
+    nic.src_ipsec_tunnels.each do |tunnel|
+      tunnel.refresh
+    end
+
+    decr_refresh_mesh
+    hop :wait
+  end
+
+  def destroy
+    DB.transaction do
+      nic.src_ipsec_tunnels_dataset.delete
+      nic.dst_ipsec_tunnels_dataset.delete
```

Comment:
Should we do any cleanup at dataplane? How is deleted ipsec tunnels cleaned up?

Reply:
Since Nics are detached/destroyed only when VM gets destroyed, we don't have to clean-up anything.